internal/controllers/helpers.go

```diff
@@ -0,0 +1,214 @@
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	yamlDecoder "k8s.io/apimachinery/pkg/util/yaml"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	utilyaml "sigs.k8s.io/cluster-api/util/yaml"
+
+	managementv3 "github.com/rancher-sandbox/rancher-turtles/internal/rancher/management/v3"
+	"github.com/rancher-sandbox/rancher-turtles/util"
+)
+
+const (
+	importLabelName           = "cluster-api.cattle.io/rancher-auto-import"
+	ownedLabelName            = "cluster-api.cattle.io/owned"
+	capiClusterOwner          = "cluster-api.cattle.io/capi-cluster-owner"
+	capiClusterOwnerNamespace = "cluster-api.cattle.io/capi-cluster-owner-ns"
+
+	defaultRequeueDuration = 1 * time.Minute
+)
+
+func getClusterRegistrationManifest(ctx context.Context, clusterName, namespace string, rancherClient client.Client,
+	insecureSkipVerify bool,
+) (string, error) {
+	log := log.FromContext(ctx)
+
+	token := &managementv3.ClusterRegistrationToken{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: namespace,
+		},
+		Spec: managementv3.ClusterRegistrationTokenSpec{
+			ClusterName: clusterName,
+		},
+	}
+	err := rancherClient.Get(ctx, client.ObjectKeyFromObject(token), token)
+
+	if client.IgnoreNotFound(err) != nil {
+		return "", fmt.Errorf("error getting registration token for cluster %s: %w", clusterName, err)
+	} else if err != nil {
+		if err := rancherClient.Create(ctx, token); err != nil {
+			return "", fmt.Errorf("failed to create cluster registration token for cluster %s: %w", clusterName, err)
+		}
+	}
+
+	if token.Status.ManifestURL == "" {
+		return "", nil
+	}
+
+	manifestData, err := downloadManifest(token.Status.ManifestURL, insecureSkipVerify)
+	if err != nil {
+		log.Error(err, "failed downloading import manifest")
+		return "", err
+	}
+
+	return manifestData, nil
+}
+
+func namespaceToCapiClusters(ctx context.Context, clusterPredicate predicate.Funcs, rancherClient client.Client) handler.MapFunc {
+	log := log.FromContext(ctx)
+
+	return func(_ context.Context, o client.Object) []ctrl.Request {
+		ns, ok := o.(*corev1.Namespace)
+		if !ok {
+			log.Error(nil, fmt.Sprintf("Expected a Namespace but got a %T", o))
+			return nil
+		}
+
+		_, autoImport := util.ShouldImport(ns, importLabelName)
+		if !autoImport {
+			log.V(2).Info("Namespace doesn't have import annotation label with a true value, skipping")
+			return nil
+		}
+
+		capiClusters := &clusterv1.ClusterList{}
+		if err := rancherClient.List(ctx, capiClusters, client.InNamespace(o.GetNamespace())); err != nil {
```

Comment:
Previously the r.Client was used here, not r.RancherClient. This is confusing, even if the invocation is using the correct r.Client.